src/hotspot/cpu/aarch64/universalNativeInvoker_aarch64.cpp

```diff
@@ -25,103 +25,303 @@
+  int _framesize;
+  OopMapSet* _oop_maps;
+public:
+  NativeInvokerGenerator(CodeBuffer* buffer,
+                         BasicType* signature,
+                         int num_args,
+                         BasicType ret_bt,
+                         const ABIDescriptor& abi,
+                         const GrowableArray<VMReg>& input_registers,
+                         const GrowableArray<VMReg>& output_registers,
+                         bool needs_return_buffer)
+   : StubCodeGenerator(buffer, PrintMethodHandleStubs),
+     _signature(signature),
+     _num_args(num_args),
+     _ret_bt(ret_bt),
+     _abi(abi),
+     _input_registers(input_registers),
+     _output_registers(output_registers),
+     _needs_return_buffer(needs_return_buffer),
+     _frame_complete(0),
+     _framesize(0),
+     _oop_maps(NULL) {
+  }
 
-  __ block_comment("init_and_alloc_stack");
+  void generate();
 
-  __ mov(Rctx, c_rarg0);
-  __ str(Rctx, Address(__ pre(sp, -2 * wordSize)));
+  int frame_complete() const {
+    return _frame_complete;
+  }
 
-  assert(_abi->_stack_alignment_bytes % 16 == 0, "stack must be 16 byte aligned");
+  int framesize() const {
+    return (_framesize >> (LogBytesPerWord - LogBytesPerInt));
+  }
 
-  __ block_comment("allocate_stack");
-  __ ldr(Rstack_size, Address(Rctx, (int) _layout->stack_args_bytes));
-  __ add(rscratch2, Rstack_size, _abi->_stack_alignment_bytes - 1);
-  __ andr(rscratch2, rscratch2, -_abi->_stack_alignment_bytes);
-  __ sub(sp, sp, rscratch2);
+  OopMapSet* oop_maps() const {
+    return _oop_maps;
+  }
+};
 
-  __ block_comment("load_arguments");
+static const int native_invoker_code_size = 1024;
 
-  __ ldr(Rsrc_ptr, Address(Rctx, (int) _layout->stack_args));
-  __ lsr(Rwords, Rstack_size, LogBytesPerWord);
-  __ mov(Rdst_ptr, sp);
+RuntimeStub* ProgrammableInvoker::make_native_invoker(BasicType* signature,
+                                                      int num_args,
+                                                      BasicType ret_bt,
+                                                      const ABIDescriptor& abi,
+                                                      const GrowableArray<VMReg>& input_registers,
+                                                      const GrowableArray<VMReg>& output_registers,
+                                                      bool needs_return_buffer) {
+  int locs_size  = 64;
+  CodeBuffer code("nep_invoker_blob", native_invoker_code_size, locs_size);
+  NativeInvokerGenerator g(&code, signature, num_args, ret_bt, abi, input_registers, output_registers, needs_return_buffer);
+  g.generate();
+  code.log_section_sizes("nep_invoker_blob");
 
-  Label Ldone, Lnext;
-  __ bind(Lnext);
-  __ cbz(Rwords, Ldone);
-  __ ldr(Rtmp, __ post(Rsrc_ptr, wordSize));
-  __ str(Rtmp, __ post(Rdst_ptr, wordSize));
-  __ sub(Rwords, Rwords, 1);
-  __ b(Lnext);
-  __ bind(Ldone);
+  RuntimeStub* stub =
+    RuntimeStub::new_runtime_stub("nep_invoker_blob",
```

Comment:
Is it acceptable for a VM fatal error to occur when the `RuntimeStub` cannot be allocated due to a (temporarily?) full code cache? If not, then you may want to do something like I'm doing in https://github.com/openjdk/jdk/pull/15970.